tardis/adapters/batchsystems/htcondor.py

```diff
@@ -4,51 +4,175 @@
 from typing import Iterable
 import logging
 
 logger = logging.getLogger("cobald.runtime.tardis.adapters.batchsystem.htcondor")
 
 
+async def htcondor_get_collectors(
+    options: AttributeDict, executor: Executor
+) -> list[str]:
+    """
+    Asynchronously retrieve a list of HTCondor collector machine names.
+
+    Runs ``condor_status -collector -af:t Machine`` (plus any additional formatted
+    options) using the provided executor, then parses the tab-delimited output
+    to extract and return the list of collector machine names.
+
+    :param options: Additional options for the ``condor_status`` call, such as
+        ``{'pool': 'htcondor.example'}``, which will be formatted and appended
+        to the command.
+    :param executor: Executor used to run the ``condor_status`` command
+        asynchronously.
+    :return: List of collector machine names.
+    """
+    class_ads = AttributeDict(Machine="Machine")
+    # Add collector query option, copy options since it is mutable
+    options = AttributeDict(collector=None, **options)
+    cmd = htcondor_status_cmd_composer(
+        attributes=class_ads,
+        options=options,
+    )
+
+    try:
+        condor_status = await executor.run_command(cmd)
+    except CommandExecutionFailure as cef:
+        logger.warning(f"condor_status could not be executed due to {cef}!")
+        raise
+
+    return [
+        row["Machine"]
+        for row in csv_parser(
+            input_csv=condor_status.stdout,
+            fieldnames=tuple(class_ads.keys()),
+            delimiter="\t",
+        )
+    ]
+
+
+async def htcondor_get_collector_start_dates(
+    options: AttributeDict, executor: Executor
+) -> dict[str, datetime]:
+    """
+    Asynchronously retrieve the master daemon start times from HTCondor for machines
+    running a collector daemon as well. Assuming both daemons have a similar start date.
+    Due to potential bug/feature in HTCondor, the DaemonStartTime of the Collector can
+    not be used directly.
+    (see https://www-auth.cs.wisc.edu/lists/htcondor-users/2025-July/msg00092.shtml)
+
+    Runs ``condor_status -master -af:t Machine DaemonStartTime`` (plus any
+    additional formatted options) using the provided executor, and parses
+    the tab-delimited output into a dictionary.
+
+    :param options: Additional options for the ``condor_status`` call, such as
+        ``{'pool': 'htcondor.example'}``, which will be formatted and appended
+        to the command.
+    :param executor: Executor used to run the ``condor_status`` command
+        asynchronously.
+    :return: List of master daemon start time for host running a collector as well.
+        (in datetime format).
+    """
+    class_ads = AttributeDict(Machine="Machine", DaemonStartTime="DaemonStartTime")
+    htcondor_collectors = await htcondor_get_collectors(options, executor)
+
+    # Add master query option, copy options since it is mutable
+    options = AttributeDict(master=None, **options)
```

Comment:
General comment: Perhaps we should implement the `|` operator for AttributeDict ([see `d | other` for `dict`](https://docs.python.org/3/library/stdtypes.html#dict.values)). This way, you could just write `options = options | {'master': None}`.

No need to make this part of this PR.